feat: add docs for searchbase (#156)
appbaseio/Docs: content/docs/reactivesearch/searchbase/overview/apireference.md

>
> This API reference is meant for the headless usage of the `Searchbase` library.

## How does it work?

The working of `Searchbase` can be better explained by the following chart.
<br/><br/>

<img alt="Dataset" src="/images/searchbase.png" />

The `Searchbase` library is a headless implementation of the core architecture in Vanilla JS, which can be used with your own UI or alongside any framework. It provides all the necessary utilities to build a powerful search UI and can be easily integrated with any UI framework. Searchbase lib maintains the search `state` and provides some `actions` which can be used to manipulate the state from the UI. It also provides events which can be used to listen for the state changes & update the UI accordingly.

> Although we don't ship any UI component with `Searchbase` directly, we provide easy to integrate libraries for different platforms. Please check `@appbaseio/react-searchbox` and `@appbaseio/vue-searchbox` for the React and Vue. The UI solution for `angular` is work in progress.

Reviewer: ```suggestion
Although we don't ship any UI component with `Searchbase` directly, we provide easy to integrate libraries for different platforms. You can check `@appbaseio/react-searchbox` and `@appbaseio/vue-searchbox` for React and Vue. A UI solution for `angular` is on our roadmap as well.
```

Reviewer: Link to the docs of React Searchbox and Vue Searchbox.

-   a search bar component,
-   a category filter component,
-   a price range component,
-   a location filter component,
-   a component to render the search results etc.

The `SearchBase` class holds the state for all the active components and can be used to provide the global configuration to the registered components, it serves the following tasks:

-   To `register` a component by unique `id`
-   To `unregister` a component by `id`
-   To retrieve the instance of the `Component` class by `id`

> Note:
>
> 1. The `id` property is a unique identifier to each search component.
> 2. The `SearchBase` class is useful when you're using multiple components that depend on each other. For example, a filter component (to display the category options) depends on the search query (search component).
>    If you're only using a single component then [Component](docs/reactivesearch/searchbase/overview/component/) class should work well.

## Constructor

The constructor of searchbase is called with the following properties:

```js
const searchbase = new SearchBase(properties);
```

### Properties

> #### To configure the Appbase environments

Reviewer: ```suggestion
#### Configure appbase.io environment
```

> The following properties can be used to configure the appbase environments globally i.e for all registered components. You can also configure those properties for each [Component](docs/reactivesearch/searchbase/overview/component/) too.

Reviewer: ```suggestion
The following properties can be used to configure appbase.io environment globally, i.e. for all registered components. You can also configure these properties for each [Component](docs/reactivesearch/searchbase/overview/component/) as well.
```

-   **index** `string` [Required]
    Refers to an index of Elasticsearch cluster.

    `Note:` Multiple indexes can be connected to by specifying comma-separated index names.

-   **url** `string` [Required]
    URL for the Elasticsearch cluster

-   **credentials** `string` [Required]

>     Basic Auth credentials if required for authentication purposes. It should be a string of the format `username:password`. If you are using an appbase.io app, you will find credentials under your [API credentials page](https://dashboard.appbase.io/app?view=credentials). If you are not using an appbase.io app, credentials may not be necessary - although having open access to your Elasticsearch cluster is not recommended.

Reviewer: API credentials page URL needs to be fixed.

-   **headers** `Object`
    set custom headers to be sent with each server request as key/value pairs. For example:

```ts
const searchbase = new Searchbase({
	index: 'gitxplore-app',
	url: 'https://@arc-cluster-appbase-demo-6pjy6z.searchbase.io',
	credentials: 'a03a1cb71321:75b6603d-9456-4a5a-af6b-a487b309eb61',
	headers: {
		secret: 'searchbase-is-awesome',
	},
});
```

-   **transformRequest** `(requestOptions: Object) => Promise<Object>`
    Enables transformation of network request before execution. This function will give you the request object as the param and expect an updated request in return, for execution.<br/>
    For example, we will add the `credentials` property in the request using `transformRequest`.
    ```js
    const searchbase = new Searchbase({
    	index: 'gitxplore-app',
    	url: 'https://@arc-cluster-appbase-demo-6pjy6z.searchbase.io',
    	credentials: 'a03a1cb71321:75b6603d-9456-4a5a-af6b-a487b309eb61',
    	transformRequest: request =>
    		Promise.resolve({
    			...request,
    			credentials: include,
    		}),
    });
    ```
-   **transformResponse** `(response: any) => Promise<any>`

>     Enables transformation of search network response before rendering them. It is an asynchronous function which will accept an Elasticsearch response object as param and is expected to return an updated response as the return value.<br/>

Reviewer: ```suggestion
    Enables transformation of search network response before rendering it. It is an asynchronous function which will accept an Elasticsearch response object as param and is expected to return an updated response as the return value.<br/>
```

```ts
const searchBase = new SearchBase({
	index: 'gitxplore-app',
	url: 'https://@arc-cluster-appbase-demo-6pjy6z.searchbase.io',
	credentials: 'a03a1cb71321:75b6603d-9456-4a5a-af6b-a487b309eb61',
});

const searchComponent = new Component({
	id: 'search-component',
	index: 'gitxplore-app',
	url: 'https://@arc-cluster-appbase-demo-6pjy6z.searchbase.io',
	credentials: 'a03a1cb71321:75b6603d-9456-4a5a-af6b-a487b309eb61',
	dataField: ['title', 'description'],
	value: '',
});

searchBase.register('search-component', searchComponent);
```

Additionally, you can override the global configurations by defining it for a particular component. For example, to register a component with a different `index` name.

```ts
const searchBase = new SearchBase({
	index: 'gitxplore-app',
	url: 'https://@arc-cluster-appbase-demo-6pjy6z.searchbase.io',
	credentials: 'a03a1cb71321:75b6603d-9456-4a5a-af6b-a487b309eb61',
});

searchBase.register('search-component', {
    index: 'gitxplore-app-2',
	dataField: ['title', 'description'],
	value: '',
});
```

#### unregister

```ts
unregister(id: string): void
```

> This method is useful to un-register a component by `id`. It is a good practice to un-register(remove) an un-mounted/un-used component to avoid the side-effects.

Reviewer: ```suggestion
This method is useful to unregister a component by `id`. It is a good practice to unregister (remove) an unmounted/unused component to avoid any side-effects.
```
